triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.oauth.middleware
+
+import akka.Done
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.Http.ServerBinding
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.HttpCookie
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.unmarshalling.{Unmarshal, Unmarshaller}
+import com.daml.oauth.server.{Request => OAuthRequest, Response => OAuthResponse}
+import com.typesafe.scalalogging.StrictLogging
+import java.util.Base64
+import java.util.UUID
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+import scala.util.Try
+import spray.json._
+
+// This is an implementation of the trigger service authentication middleware
+// for OAuth2 as specified in `/triggers/service/authentication.md`
+object Server extends StrictLogging {
+  import com.daml.oauth.server.JsonProtocol._
+
+  private def toRedirectUri(uri: Uri) =
+    Uri()
+      .withScheme(uri.scheme)
+      .withAuthority(uri.authority)
+      .withPath(Uri.Path./("cb"))
+
+  def start(
+      config: Config)(implicit system: ActorSystem, ec: ExecutionContext): Future[ServerBinding] = {
+    implicit val unmarshal: Unmarshaller[String, Uri] = Unmarshaller.strict(Uri(_))
+    // TODO[AH] Make sure this is bounded in size - or avoid state altogether.
+    val requests = TrieMap[UUID, Uri]()
+    val route = concat(
+      path("auth") {
+        get {
+          complete((StatusCodes.NotImplemented, "The /auth endpoint is not implemented yet"))
+        }
+      },
+      path("login") {
+        get {
+          parameters(('redirect_uri.as[Uri], 'claims))
```

Comment:
Maybe worth extracting the request bodies into separate functions? Totally fine to leave that for separate PRs.

Reply:
Yeah, that's probably better for readability, especially once the other endpoints are filled in.